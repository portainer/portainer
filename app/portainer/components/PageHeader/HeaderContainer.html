<div class="row header">
  <div id="loadingbar-placeholder"></div>
  <div class="col-xs-12"> <div class="meta" ng-transclude></div></div>
</div>
