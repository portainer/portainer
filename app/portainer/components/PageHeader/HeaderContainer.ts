export const Header = {
  transclude: true,
  templateUrl: './HeaderContainer.html',
};
