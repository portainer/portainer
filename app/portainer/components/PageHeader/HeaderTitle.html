<div class="page white-space-normal">
  {{ $ctrl.titleText }}
  <span class="header_title_content" ng-transclude></span>
  <span class="pull-right user-box" ng-if="$ctrl.username && $ctrl.display">
    <i class="fa fa-user-circle" aria-hidden="true"></i>
    {{ $ctrl.username }}
  </span>
</div>
