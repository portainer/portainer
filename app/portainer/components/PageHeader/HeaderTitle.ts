import controller from './HeaderTitle.controller';

export const HeaderTitle = {
  requires: '^rdHeader',
  bindings: {
    titleText: '@',
  },
  transclude: true,
  templateUrl: './HeaderTitle.html',
  controller,
};
