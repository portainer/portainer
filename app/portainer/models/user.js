export function UserViewModel(data) {
  this.Id = data.Id;
  this.Username = data.Username;
  this.Role = data.Role;
  this.UserTheme = data.UserTheme;
  if (data.Role === 1) {
    this.RoleName = 'administrator';
  } else {
    this.RoleName = 'user';
  }
  this.AuthenticationMethod = data.AuthenticationMethod;
  this.Checked = false;
  this.EndpointAuthorizations = null;
  this.PortainerAuthorizations = null;
}

export function UserTokenModel(data) {
  this.id = data.id;
  this.userId = data.userId;
  this.description = data.description;
  this.prefix = data.prefix;
  this.dateCreated = data.dateCreated;
  this.lastUsed = data.lastUsed;
}
