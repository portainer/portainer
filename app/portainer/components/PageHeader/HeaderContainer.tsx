import { PropsWithChildren, createContext, useContext } from 'react';

import './HeaderContainer.css';

const Context = createContext<null | boolean>(null);

export function useHeaderContext() {
  const context = useContext(Context);

  if (context == null) {
    throw new Error('Should be nested inside a HeaderContainer component');
  }
}

export function HeaderContainer({ children }: PropsWithChildren<unknown>) {
  return (
    <Context.Provider value>
      <div className="row header">
        <div id="loadingbar-placeholder" />
        <div className="col-xs-12">
          <div className="meta">{children}</div>
        </div>
      </div>
    </Context.Provider>
  );
}

export const HeaderAngular = {
  transclude: true,
  templateUrl: './HeaderContainer.html',
};
