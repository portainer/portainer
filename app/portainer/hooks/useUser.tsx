import jwtDecode from 'jwt-decode';
import { useCurrentStateAndParams } from '@uirouter/react';
import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState,
  useMemo,
} from 'react';

import { getUser } from '../users/user.service';
import { User, UserId } from '../users/types';

import { useLocalStorage } from './useLocalStorage';

interface State {
  user?: User;
}

export const UserContext = createContext<State | null>(null);

export function useUser() {
  const context = useContext(UserContext);

  if (context === null) {
    throw new Error('should be nested under UserProvider');
  }

  return useMemo(
    () => ({ user: context.user, isAdmin: isAdmin(context.user) }),
    [context.user]
  );
}

export function useAuthorizations(
  authorizations: string | string[],
  adminOnlyCE = false
) {
  const authorizationsArray =
    typeof authorizations === 'string' ? [authorizations] : authorizations;

  const { user } = useUser();
  const { params } = useCurrentStateAndParams();

  if (!user) {
    return false;
  }

  if (process.env.PORTAINER_EDITION === 'CE') {
    return !adminOnlyCE || isAdmin(user);
  }

  const { endpointId } = params;
  if (!endpointId) {
    return false;
  }

  if (isAdmin(user)) {
    return true;
  }

  if (
    !user.EndpointAuthorizations ||
    !user.EndpointAuthorizations[endpointId]
  ) {
    return false;
  }

  const userEndpointAuthorizations = user.EndpointAuthorizations[endpointId];
  return authorizationsArray.some(
    (authorization) => userEndpointAuthorizations[authorization]
  );
}

interface AuthorizedProps {
  authorizations: string | string[];
  children: ReactNode;
}

export function Authorized({ authorizations, children }: AuthorizedProps) {
  const isAllowed = useAuthorizations(authorizations);

  return isAllowed ? <>{children}</> : null;
}

interface UserProviderProps {
  children: ReactNode;
}

export function UserProvider({ children }: UserProviderProps) {
  const [jwt] = useLocalStorage('JWT', '');
  const [user, setUser] = useState<User | undefined>();

  useEffect(() => {
    if (jwt !== '') {
      const tokenPayload = jwtDecode(jwt) as { id: number };

      loadUser(tokenPayload.id);
    }
  }, [jwt]);

  const providerState = useMemo(() => ({ user }), [user]);

  if (jwt === '') {
    return null;
  }

  if (providerState.user === null) {
    return null;
  }

  return (
    <UserContext.Provider value={providerState}>
      {children}
    </UserContext.Provider>
  );

  async function loadUser(id: UserId) {
    const user = await getUser(id);
    setUser(user);
  }
}

function isAdmin(user?: User): boolean {
  return !!user && user.Role === 1;
}

export function useIsAdmin() {
  const { user } = useUser();
  return !!user && isAdmin(user);
}
