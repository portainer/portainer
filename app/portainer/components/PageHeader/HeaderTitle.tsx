import { PropsWithChildren } from 'react';

import { useUser } from '@/portainer/hooks/useUser';

import { useHeaderContext } from './HeaderContainer';
import controller from './HeaderTitle.controller';

interface Props {
  title: string;
}

export function HeaderTitle({ title, children }: PropsWithChildren<Props>) {
  useHeaderContext();
  const { user } = useUser();

  return (
    <div className="page white-space-normal">
      {title}
      <span className="header_title_content">{children}</span>
      {user && !window.ddExtension && (
        <span className="pull-right user-box">
          <i className="fa fa-user-circle" aria-hidden="true" />
          {user.Username}
        </span>
      )}
    </div>
  );
}

export const HeaderTitleAngular = {
  requires: '^rdHeader',
  bindings: {
    titleText: '@',
  },
  transclude: true,
  templateUrl: './HeaderTitle.html',
  controller,
};
