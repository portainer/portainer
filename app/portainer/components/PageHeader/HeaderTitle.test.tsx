import { UserContext } from '@/portainer/hooks/useUser';
import { UserViewModel } from '@/portainer/models/user';
import { render } from '@/react-tools/test-utils';

import { HeaderContainer } from './HeaderContainer';
import { HeaderTitle } from './HeaderTitle';

test('should not render without a wrapping HeaderContainer', async () => {
  const consoleErrorFn = jest
    .spyOn(console, 'error')
    .mockImplementation(() => jest.fn());

  const title = 'title';
  function renderComponent() {
    return render(<HeaderTitle title={title} />);
  }

  expect(renderComponent).toThrowErrorMatchingSnapshot();

  consoleErrorFn.mockRestore();
});

test('should display a HeaderTitle', async () => {
  const username = 'username';
  const user = new UserViewModel({ Username: username });

  const title = 'title';
  const { queryByText } = render(
    <UserContext.Provider value={{ user }}>
      <HeaderContainer>
        <HeaderTitle title={title} />
      </HeaderContainer>
    </UserContext.Provider>
  );

  const heading = queryByText(title);
  expect(heading).toBeVisible();

  expect(queryByText(username)).toBeVisible();
});
