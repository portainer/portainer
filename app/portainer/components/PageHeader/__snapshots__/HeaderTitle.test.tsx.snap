// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not render without a wrapping HeaderContainer 1`] = `"Should be nested inside a HeaderContainer component"`;
